### DISCUSSION TITLE
User account synchronization in Microservices architecture

### DISCUSSION THREAD

**Discussion by dberardo** [keycloak/keycloak#16056] [Q&A]
I know keycloak as a IAM provider to perform SSO in a standardized way among different web applications, mainly using OIDC protocol.

Now, i am try setting up a system whereby some of the Microservices involved needs to constantly synchronize information related to user accounts, e.g. their email address, phone number, permissions, groups they belong to, etc.

LDAP could be used for that, but afaik Keycloak does not provide an LDAP server API and so it is not possible to query user information periodically. 

I am not aware of any OIDC-specifoc way to achieve such a use case, especially when the information that is required about the user by one such Microservices is just too big to fit a single JWT.

Do you have any recommendations on how to setup such a user account synchronization system?

**Comment by mzndr** [comment]
I know this is old, but in my case, we're using RabbitMQ as a message bus for inter-service communication. Theres [this plugin](https://github.com/aznamier/keycloak-event-listener-rabbitmq) to publish events, like user registration to the message bus. I'm listening to these events in a service to get the ID of the user from the event, and then query the user via Keycloaks REST-API. Then the user gets transformed and published on the bus for other services to synchronize. [There is also a plugin to publish events via HTTP to an endpoint](https://github.com/p2-inc/keycloak-events#http-sender), which probably is useful for people that don't use RabbitMQ. I'm actually looking for a cleaner, less hacky feeling solution right now.

**Comment by danielFesenmeyer** [comment]
From my experience, it is quite complicated to synchronize users based on events - because

- there are so many different events to be tracked
- there is no unique way to figure out which user is affected by an event (you have to parse URIs for Admin API changes)
- some changes currently don't provide an event at all (e.g. changes by IDP mappers)

Probably it is easier to track changes on database level, with a tool like debezium: https://debezium.io/.
Of course, your synchronization is then tightly coupled to the database schema.

**Reply by mzndr** [reply]
I think the event thing works for our simple use case for now, but as I said, I can already tell that its quite shaky. Doing it on database level came to mind but I didn't look into it, thanks for pointing me out to debezium.

**Reply by danielFesenmeyer** [reply]
I agree, for simple use cases the event approach works fine. We currently still use that approach, but with growing complexity (new use cases) it definitely got more involved.

Btw, it was @thomasdarimont who suggested debezium in the context of a somewhat related discussion on the dev mailing list: https://groups.google.com/g/keycloak-dev/c/s1O7SeWywiA

**Reply by dberardo-com** [reply]
thanks for the answers you provided. indeed, if i had to do event-based synchronization i would opt for kafka/debezium, as that seemed to work quite well for other applications. however, i am trying to use a "pull-based" (queries) snychronization, meaning that the microservices request information to keycloak only when needed. thats why i was thinking at using LDAP. LDAP is not great, but it is quite standard and used in a lot of platforms (think of Gitlab) to synchronize user information from active directory and others ... thats why i wonder whether such thing exists on keycloak as well

**Comment by slominskir** [comment]
You can use the Admin API to do lookups as needed from the client:

REST API:
https://www.keycloak.org/docs-api/21.0.0/rest-api/

Java API:
https://javadoc.io/doc/org.keycloak/keycloak-admin-client/latest/index.html

I've done this in a Wildfly app, and I added a simple in-memory caching layer to avoid hammering Keycloak.   In this particular app I can tolerate stale profile info (cache clears every 24 hours, else I can clear it manually if I modify a group/role and user is impatiently waiting) as it almost never changes and user load is relatively light (and small).   If your use case differs significantly than a pub/sub message bus like Kafka may be preferrable to keep a local database cache more up-to-date (though that's an enormous amount of architecture to maintain if you don't use it already).  Or just forgo caching layer and scale up Keycloak with another host.   Or possibly use the same database for Keycloak and your apps and use cross schema links / synonyms to read directly from Keycloak schema.

**Comment by artemrogov** [comment]
Hello everyone) I use the RabbitMQ broker as a broker in my microservices; in my case, events are listened to - updating, creating and deleting a profile. It works. I wrote a plugin for keylock that sends messages to a topic that are listened to by keycloak, the consumer of the microservice listens to the topic and receives new users, in this case he receives a unique inter-service identifier and some registration data - attributes. I don't know how true this is, but it works.

**Reply by dberardo-com** [reply]
> topic that are listened to by keycloak

i assume this is possible thanks to the custom plugin you wrote, isnt it ? is it publicly available ?
